Feature/haeun step1

pear96 opened the pull request:
상세 내용은 각 커밋 메세지에 자세히 적어두었습니다.
PR에는 고민점을 동일하게 작성해 두겠습니다.

[FileMetaData]
- fileId에 대해 Getter로 사용될 일이 없는데, fileId만 제외한 다른 필드에 다 @getter를 붙일지, 아니면 클래스에 @getter를 붙인 것을 유지할지 고민입니다.

[FileRepository]
- 파일 로컬 저장 이름이 UUID로 약간 길어졌는데, 이로 인해 검색 성능에 큰 악영향을 주는지 궁금합니다.
  파일 로컬 저장 이름 VS 파일 원본 이름 + 사용자 이름 중에 어느 쿼리가 나을지 고민됩니다.
  (대신 이를 위해 DB에 한 사용자의 같은 파일명에 대해 (1), (2)등 넘버링으로 구분해서 저장해야함...)

[FileException]
- CustomException에 대해 메세지를 포함할까 고민이 됩니다. 
- 메세지를 포함하지 않은 이유는 FileService에서 예외를 던질 때 직접 메세지를 작성하고 싶지 않았기 때문입니다.
- 메세지에 대해 enum을 만들까요..?

[ApiExceptionRes]
- 예외 발생시 클라이언트에 응답을 전달하는 DTO입니다.
- FileException에서 메세지를 전달하지 않기 때문에, 마찬가지로 메세지를 포함할지 말지 고민이 됩니다.
  만약 화면단이 있다면, 오류 응답에 대해 메세지가 일치했을 때 처리하기보단, 응답 코드에 따라 화면을 바꿀 것이라고 생각하기 때문입니다.

[ExceptionHandlingController]
- 컨트롤러에서 발생한 예외를 처리하는 클래스를 생성했습니다.
- ConstraintViolatinException은 클라이언트가 전달한 값이 조건을 만족시키지 못할 때 발생합니다.
- MissingServletRequestPartException은 클라이언트가 아예 어떠한 값을 전송하지 않았을 때 발생합니다.
- 실제라면 Missing~ 예외는 발생하지 않겠죠..?

[FileControllerTest]
- 처음에 FileService에 대해 테스트를 만들다가, 컨트롤러 테스트로 수정하였습니다. 컨트롤러가 맞을까요?

pear96 committed:
chore: 프로젝트 설정

- MySQL을 사용하기 위해 connector, JPA를 추가
- 유효성 검사를 위해 Validation 패키지 추가

pear96 committed:
feat: 파일 업로드/삭제/다운로드 구현

[FileMetaData]
- 로컬에 저장될 파일 이름과 실제 파일 이름을 구분지어 저장했습니다.
- fileId에 대해 Getter로 사용될 일이 없는데, fileId만 제외한 다른 필드에 다 @Getter를 붙일지,
  아니면 클래스에 @Getter를 붙인 것을 유지할지 고민입니다.

[FileController]
- uploadFile : MultiPartFile을 받기 위해 RequestPart를 사용했습니다. (참고 : https://middleearth.tistory.com/35)
- 입력 파라미터에 유효성을 검사하도록 어노테이션을 활용했습니다.
  - MultiPartFile은 Null검사를 Spring에서 자동으로 하기 때문에 따로 어노테이션을 사용하지 않았습니다.

[FileService]
- uploadFile : file이 비어있는 경우를 서비스 안에서 처리했습니다.
  MultiPartfile에 @NotEmpty를 붙일 수 없기 때문입니다.
- 파일명 중복을 방지하기 위해 UUID를 파일명 앞에 붙였습니다.
- 파일 삭제, 파일 다운로드에서 파일을 가져오는 로직이 동일해 getFile 메서드를 생성했습니다.
  - getFile : 파일을 받을 수 없는 경우, FileMetaData를 삭제/다운로드 메서드에 반환할 수 없게 됩니다.
  - 따라서 FileException을 따로 생성하여 예외를 처리했습니다.
  - 전체적으로 예외 처리 로직을 통일하고자 ResponseEntity.bad_request()를 반환하는 대신 Exception을 활용했습니다.

[FileRepository]
- 파일의 로컬 저장 이름으로 파일 메타데이터를 조회하는 메서드 생성
- 파일 로컬 저장 이름이 UUID로 약간 길어졌는데, 이로 인해 검색 성능에 큰 악영향을 주는지 궁금합니다.
  파일 로컬 저장 이름 VS 파일 원본 이름 + 사용자 이름 중에 어느 쿼리가 나을지 고민됩니다.
  (대신 이를 위해 DB에 한 사용자의 같은 파일명에 대해 (1), (2)등 넘버링으로 구분해서 저장해야함...)

pear96 committed:
feat: 예외 처리

[FileException]
- 파일 서비스 과정에서 발생하는 예외를 처리하는 CustomException을 만들었습니다.
- 메세지를 포함할까 고민이 됩니다. 메세지를 포함하지 않은 이유는 예외를 던질 때 직접 메세지를 작성하고 싶지 않았기 때문입니다.
- 메세지에 대해 enum을 만들까요..?

[ApiExceptionRes]
- 예외 발생 시 클라이언트에 반환하는 응답 DTO 입니다.
- 마찬가지로 메세지를 포함할지 말지 고민이 됩니다.
  만약 화면단이 있다면, 오류 응답에 대해 메세지가 일치했을 때 처리하기보단, 응답 코드에 따라 화면을 바꾸기 때문입니다.

[ExceptionHandlingController]
- 컨트롤러에서 발생한 예외를 처리하는 클래스를 생성했습니다.
- ConstraintViolatinException은 클라이언트가 전달한 값이 조건을 만족시키지 못할 때 발생합니다.
- MissingServletRequestPartException은 클라이언트가 아예 어떠한 값을 전송하지 않았을 때 발생합니다.
- 실제라면 Missing~ 예외는 발생하지 않겠죠..?

pear96 committed:
test: 파일 컨트롤러 테스트 코드 작성

테스트 목록은 다음과 같습니다.
(테스트 코드를 처음 작성해보는거라.. 피드백 주시면 감사드립니다.)

- 파일 업로드
  - 성공
  - 실패
    1. 사용자 이름 없음
    2. 파일 없음
      - 파일이 아예 전달되지 않는 상황을 구현하기 위해 MockMvc를 활용했습니다.

- 파일 삭제
  - 성공
  - 실패
    1. 삭제할 파일 이름 없음
    2. 사용자 이름 없음
    3. 사용자 이름이 불일치

- 파일 다운로드
  - 성공
  - 실패
    1. 파일 이름 없음
    2. 사용자 이름 없음
    3. 사용자 이름이 불일치

pear96 committed:
refactor: 파일 업로드 NPE 처리 및 테스트 코드스멜 개선

- 파일 업로드시 @NotEmpty는 안되더라도 @NotNull은 처리할 수 있도록 했습니다.
- 테스트 코드에서 파일 삭제 / 다운로드시 파일 업로드를 미리 진행하는데,
  이때 200 응답 코드와 응답 결과의 NotNull을 처리하는 코드를 추가했습니다.

pear96 commented:
application-local.yml이 없어서 빌드에 실패했습니다.
DB정보 application-local.yml에 따로 저장해뒀는데, 그냥 올려버려야 하나요?
제 로컬 DB가 털릴 일이 없을 것 같긴 한데..

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/repository/FileRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface FileRepository extends JpaRepository<FileMetaData, Long> {
```
Jpa 레포지토리를 타고 올라가면 어떤 인터페이스가 있나요!?

VSFe replied:
그러게요

pear96 replied:
Repository<T, ID> : 중앙 저장소 마커 인터페이스
- PagingAndSortingRepository<T, ID> : 엔티디들을 페이지네이션, 정렬하는 레포지토리
  - ListPagingAndSortingRepository<T, ID> : List에 담아 반환해줌
- CrudRepository<T, ID> : save, find, delete, count 등 CRUD 수행
  - ListCrudRepository<T, ID> : saveAll, findAll 의 결과를 List에 담아 반환

QueryByExampleExecutor<T> : 예제(찾고싶은 값을 가진Entity)객체를 생성해서, 이와 일치하는 엔티티 반환

등이 있군요...

kochungcheon replied:
> * istPagingAndSortingRepository<T, ID> : List에 담아 반환해줌
> * CrudRepository<T, ID> : save, find, delete, count 등 CRUD 수행
>   
>   * ListCrudRepository<T, ID> : saveAll, findAll 의 결과를 List에 담아 반환
> 
> QueryByExampleExecutor : 예제(찾고싶은 값을 가진Entity)객체를 생성해서, 이와 일치하는 엔티티 반환
> 
> 등이 있군요...
그렇습니다 ! extends JpaRepository 한 순간부터 레포지토리 빈으로 등록 되기 떄문에 @ Repository 는 빼셔도 될 거같아요!
https://stackoverflow.com/questions/51918181/why-isnt-necessary-repository-for-this-spring-boot-web-app

pear96 committed:
chore: yml 정보 서브모듈 설정

- yml을 분리하자 Build에 실패하여 서브모듈의 정보를 읽도록 수정했습니다.

kochungcheon commented on src/test/java/com/c4cometrue/mystorage/controller/FileControllerTest.java:
```diff
@@ -0,0 +1,307 @@
+    @DisplayName("파일 업로드 성공")
+    void uploadFileSuccess() {
+        // 요청 payload
+        MultipartFile multipartFile = new MockMultipartFile(
```
테스트에 자주 사용되는 변수들은 어떤 식으로 관리하면 좋을 까요!?

pear96 replied:
저도 청천님처럼 따로 클래스에 static 변수로 만드려고요!

kochungcheon reviewed (commented):
고생하셨습니다!😀

VSFe commented on src/main/java/com/c4cometrue/mystorage/controller/FileController.java:
```diff
@@ -0,0 +1,40 @@
+
+    private final FileService fileService;
+
+    @PostMapping("")
```
""은 생략 가능합니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/controller/FileController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @DeleteMapping("")
+    public ResponseEntity<String> deleteFile(@NotBlank @RequestParam("filename") String fileStorageName,
```
기왕이면 `@RequestParam("~~~")` 을 쓰는 것 보단, 외부에서 받는 이름과 내부에서 사용하는 변수의 이름을 통일시키고 제거하는게 더 혼선이 적을 것 같아요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/dto/response/CreateFileRes.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Setter;
+
+@Setter
+@Getter
```
@Setter 를 사용하지 않으면, 과감하게 제거해주세요.

Java의 트랜드가 FP와 Switch Pattern Matching으로 변화하면서, **불변 클래스**의 중요도가 좀 더 높아졌어요.
클래스를 불변으로 두게 되면 활용할 수 있는 범위가 넓어지기도 하고, 기본적으로 DTO는 불변임이 전제된 상태에서 개발을 많이 하는 편이에요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/dto/response/CreateFileRes.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Getter;
+import lombok.Setter;
+
+@Setter
```
추가적으로, Java 17을 쓰고 있기 때문에, 불변 클래스를 사용한다면 `Record`로 변경해서 사용하는 것을 권장합니다.

참고: https://s7won.tistory.com/2

VSFe commented on src/main/java/com/c4cometrue/mystorage/entity/FileMetaData.java:
```diff
@@ -0,0 +1,44 @@
+@Getter
+public class FileMetaData {
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int fileId;
```
일반적으로 PK는 long으로 많이 사용해요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    public FileMetaData getFile(String fileStorageName, String username){
+        FileMetaData fileMetaData = fileRepository.findByFileStorageName(fileStorageName);
```
`optional`로 받아서 elseThrow로 처리하는게 좀 더 자연스러울 것 같아요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,110 @@
+        Path filePath = storagePath.resolve(fileMetaData.getFileStorageName());
+        try {
+            // 실제 파일과 파일 데이터 삭제
+            Files.delete(filePath);
```
물리적인 파일 삭제 -> DB 삭제
DB 삭제 -> 물리적인 파일 삭제

어떤 순서로 접근하는게 맞을까요? 왜 그럴까요?

pear96 replied:
사용자 위주로 생각하면 실제 물리 파일은 서버에 남아있어도 DB에는 파일이 삭제되었다고 뜨는게 안전할 것 같은데.. (없는 파일에 대해 삭제, 다운로드 요청을 보낼 수 없으니까)
서버 저장소 위주로 생각하면 DB엔 없는 파일들이 저장소의 공간을 낭비하고 있을때, 이 파일들에 대해 DB에 없을 경우 삭제하는 배치를 따로 돌려야하고.. 근데 저장소에 파일이 무지 많다면 이것도 부담스럽겠네요..

차라리 DB에 있고 물리적 파일이 없을때, 해당 파일에 대해 요청이 오고 이때 파일을 못 찾는다면 DB에서 삭제하고 Not found를 보내야하나..??? 

근데 안정성을 더 중시한다면 역시 DB 삭제가 먼저 수행되는 것이 맞을 것 같습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ExceptionHandlingController.java:
```diff
@@ -0,0 +1,42 @@
+@RestControllerAdvice
+public class ExceptionHandlingController {
+    // value에 포함된 예외들 처리함
+    @ExceptionHandler(value = {FileException.class})
```
선언하는 Exception의 종류가 너무 많아질 위험이 있습니다.
Exception을 하나로 묶어버리고, 내부에서 우리가 어떤 문제가 있는지 구분할 수 있도록 만들 순 없을까요?

(요건 감이 안 오면 그냥 물어보세요. 힌트 드림)

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ApiExceptionRes.java:
```diff
@@ -0,0 +1,21 @@
+import java.time.ZonedDateTime;
+
+@Getter
+public class ApiExceptionRes {
```
시연한 것 처럼, 최소한의 파일을 사용하여 발생할 수 있는 서비스적인 에러를 관리할 수 있게 해 주세요.

+ Resolver는 안 보여줬으니까, 알아서 파이팅!

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,110 @@
+                    .build();
+            // DB 저장
+            fileRepository.save(fileMetaData);
+            return ResponseEntity.ok(new CreateFileRes(fileMetaData));
```
ResponseEntity로의 반환은 Controller 단에서 수행하는게 좋습니다.
비즈니스 로직에서는 순수한 CreateFileRes를 반환하게 해 주세요.

(+ Created - HttpStatus 201)
(++ 굳이 ResponseEntity 쓰지 말고, `@ResponseStatus(HttpStatus.CREATED)`를 Controller method에 삽입해도 됩니다.)

VSFe commented on src/test/java/com/c4cometrue/mystorage/controller/FileControllerTest.java:
```diff
@@ -0,0 +1,307 @@
+@SpringBootTest
+@Transactional
+@AutoConfigureMockMvc
+class FileControllerTest {
```
통합테스트를 사용하고 있네요.
전체적인 흐름을 잡기 위해 통합테스트를 사용하고 있다면, 아예 클래스 이름을 분리해서 쓰는 것을 권장합니다.

일반적으로 클래스 이름으로 테스트를 만들게 되면, 이는 해당 클래스 자체의 메서드가 정상적으로 동작하는지를 확인하기 위해 수행하는 것이라서, 대부분 유닛테스트로 진행합니다. (외부 영향을 받는 것들은 철저하게 Mocking 하면서)

VSFe commented on src/test/java/com/c4cometrue/mystorage/controller/FileControllerTest.java:
```diff
@@ -0,0 +1,307 @@
+        );
+
+        // 파일 업로드 실패
+        assertThrows(ConstraintViolationException.class,
```
TMI인데, assertThrows의 반환값은 반환된 그 예외 자체입니다.
상황에 따라서 해당 값을 받아서 추가로 검증을 해야할 수도 있어요.

VSFe reviewed (changes requested):
수고하셨습니다~~
이후 폴더를 고려하면, 현재 구조에서 유지되어야 할 점과, 변경되어야 할 수도 있는 점이 있을텐데, 그 부분도 잘 생각해보면 좋을 것 같습니다~~~~

pear96 committed:
style: CRLF를 LF로 변경

- 모든 파일의 CRLF를 LF로 변경했습니다.
- application.yml에 내용을 전부 추가했습니다.

Commit:
chore: update checkstyle version

pear96 committed:
Merge branch 'main' into feature/haeun_step1

pear96 committed:
test: Files, Repository Mock 설정

- 실제 테스트에서 실행하기 어려운 Files 클래스의 메서드와 Repository를 Mocking했습니다.

VSFe committed:
feat: 짬통

VSFe committed:
feat: 짬통2

pear96 committed:
chore: 서브모듈 삭제

- 서브모듈 삭제

pear96 committed:
refactor: DTO 클래스 record로 바꿈

- Record 타입을 활용했습니다.
- 불변 클래스를 쉽고 간편하게 생성할 수 있습니다.
- 모든 필드가 자동으로 private final이 됩니다.
- Getter가 자동으로 구현되어 .getValue() 대신 .Value()로 조회할 수 있습니다.
- 생성자를 작성하지 않아도 되고 toString, equals, hashCode 메소드가 자동으로 구현됩니다.

pear96 committed:
refactor: Entity 컬럼명 변경

- FileMetaData의 `owner`를 `username`으로 통일했습니다.
- FileRepository에서 파일명과 유저명으로 파일을 찾는 메서드 추가
- 주석을 추가했습니다.

pear96 committed:
style: checkStyle에 맞춰 코드 스타일 수정

pear96 committed:
refactor: 예외 처리

[ErrorCd]
- ServiceException 클래스와 통일 시키기 위해 Throwable을 받는 생성자 작성
  (Throwable과 함께 서비스 레이어에서 발생시키는 경우가 없는데 필요한지 의문)

[ExceptionHandlingController]
- 디버깅 메세지를 전달하지 않도록 검증을 추가했습니다.
- ConstraintViolationException과 MissingServletRequestPartException을 처리하는 메서드를 분리하지 않고, ServiceException과 합친 다음에  Instance Of로 확인해야할지.. 아니면 ServiceException 외의 모든 예외를 한번에 처리할지 고민됩니다.

pear96 committed:
style: check style에 맞춰 변경 및 주석 추가

pear96 committed:
refactor: Resource 처리 위치 변경

[FileUtil & FileService]
- 기존의 `new UrlResource(uri)` 로는 Mocking이 불가능했습니다. 따라서 ResourceLoader를 사용하기로 했습니다.
- 그러나 FileUtil은 `Files`로 인해 static Method를 만든 상태라 ResourceLoader를 static으로 의존하기 어려웠습니다.
- 따라서 FileService에서 ResourceLoader를 의존하고 파일을 다운로드 하는 과정을 처리했습니다.

pear96 committed:
test: 테스트 코드 작성

[FileServiceTest]
- FileUtil을 static으로 Mocking하여 `Files`와 관련된 static method를 처리합니다.
- FileRepository를 mocking하여 DB 관련 로직을 처리합니다.
- 실패했을 경우에는 예상한 예외와 일치하는지, 성공했다면 정해진 메서드가 정상적으로 실행되었는지를 검증합니다.

[FileUtilTest]
- `Files` 를 static mocking하고, 정해진 상황에서 `Files.copy()` 또는 `Files.delete()`가 제대로 실행되는지 확인합니다.
- 실패한 경우 예상한 예외코드와 일치하는지 검증합니다.

[FileControllerTest]
- 파일 업로드 / 삭제 / 다운로드에 대해 각각 실제로 fileService의 메서드가 실행되었는지 검증했습니다.

[ErrorCdTest]
- ServiceException 생성자에 대해 정상적으로 생성하는지 확인합니다.
- 그 외 ENUM값으로 생성되어있는 각 예외에 대해 에러 코드와 메세지를 검증합니다.

[ExceptionHandlingControllerTest]
- 예외 발생시 예상 에러 코드와 메세지를 반환하는지 확인합니다.
- 디버그 메세지가 있는 경우도 검증합니다.
- ValidException에 대해 처리했습니다.

[StoragePathServiceTest]
- Path를 제대로 생성하는지 검증했습니다.

[FileMetaData, ApiExceptionRes Test]
- 생성자 테스트를 검증했습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ExceptionHandlingController.java:
```diff
@@ -0,0 +1,51 @@
+    @ExceptionHandler(value = {ServiceException.class})
+    public ResponseEntity<ApiExceptionRes> handleServiceException(ServiceException exception) {
+        if (exception.getDebugMessage() != null) {
+            logger.debug(exception.getMessage());
```
👍

VSFe commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,17 @@
+    password: pear00!
+
+storage:
+  dir: C:/Users/pear/study/C4/storage
```
이후에 배포 고려했을 땐, profile로 분기해서 처리할 수 있도록 해주세요~
(지금 말고)

pear96 replied:
넵 다음부터는 청천님처럼 코멘트 달겠습니다~

VSFe reviewed (approved):
(코드는 제가 라이브로 수정한게 좀 있으니 따로 코멘트 달라고 하진 않을게요. 하지만 다음 스텝부터는 리뷰에 대해서 청천님 PR 처럼 코멘트 부탁드릴게요.)

Code Smell만 나중에 수정해주세요~
